program/tests/harvest_rewards.rs

```diff
@@ -0,0 +1,557 @@
+    setup_mint(&mut context, &mint, &Pubkey::new_unique(), 0).await;
+
+    let instruction = harvest_rewards(&holder_rewards_pool, &holder_rewards, &token_account, &mint);
+
+    let transaction = Transaction::new_signed_with_payer(
+        &[instruction],
+        Some(&context.payer.pubkey()),
+        &[&context.payer],
+        context.last_blockhash,
+    );
+
+    let err = context
+        .banks_client
+        .process_transaction(transaction)
+        .await
+        .unwrap_err()
+        .unwrap();
+
+    assert_eq!(
+        err,
+        TransactionError::InstructionError(
+            0,
+            InstructionError::Custom(PaladinRewardsError::IncorrectHolderRewardsAddress as u32)
+        )
+    );
+}
+
+#[tokio::test]
+async fn fail_holder_rewards_invalid_data() {
+    let owner = Pubkey::new_unique();
+    let mint = Pubkey::new_unique();
+
+    let token_account = get_associated_token_address(&owner, &mint);
+    let holder_rewards = get_holder_rewards_address(&token_account);
+    let holder_rewards_pool = get_holder_rewards_pool_address(&mint);
+
+    let mut context = setup().start_with_context().await;
+    setup_holder_rewards_pool_account(&mut context, &holder_rewards_pool, 0, 0).await;
+    setup_token_account(&mut context, &token_account, &owner, &mint, 0).await;
+    setup_mint(&mut context, &mint, &Pubkey::new_unique(), 0).await;
+
+    // Setup holder rewards account with invalid data.
+    {
+        context.set_account(
+            &holder_rewards,
+            &AccountSharedData::new_data(
+                100_000_000,
+                &vec![5; 165],
+                &paladin_rewards_program::id(),
+            )
+            .unwrap(),
+        );
+    }
+
+    let instruction = harvest_rewards(&holder_rewards_pool, &holder_rewards, &token_account, &mint);
+
+    let transaction = Transaction::new_signed_with_payer(
+        &[instruction],
+        Some(&context.payer.pubkey()),
+        &[&context.payer],
+        context.last_blockhash,
+    );
+
+    let err = context
+        .banks_client
+        .process_transaction(transaction)
+        .await
+        .unwrap_err()
+        .unwrap();
+
+    assert_eq!(
+        err,
+        TransactionError::InstructionError(0, InstructionError::InvalidAccountData)
+    );
+}
+
+#[allow(clippy::arithmetic_side_effects)]
+#[test_case(0, 0, 0, 0)]
+#[test_case(10, 5, 10, 0)]
+#[test_case(10, 5, 10, 2)]
+#[test_case(10, 5, 10, 5)]
+#[test_case(500_000, 1_000, 50_000_000, 0)]
+#[test_case(500_000, 1_000, 50_000_000, 50_000)]
+#[test_case(500_000, 1_000, 50_000_000, 100_000)]
+#[test_case(100_000_000, 10_000_000, 50_000_000, 0)]
+#[test_case(100_000_000, 10_000_000, 50_000_000, 2_500_000)]
+#[test_case(100_000_000, 10_000_000, 50_000_000, 5_000_000)]
+#[test_case(20_000_000_000, 50_000_000, 40_000_000_000, 0)]
+#[test_case(20_000_000_000, 50_000_000, 40_000_000_000, 50_000_000)]
+#[test_case(20_000_000_000, 50_000_000, 40_000_000_000, 100_000_000)]
```

Comment:
Can you add another test-case variable for the amount of lamports in the rewards pool account?

Reply:
Yep, I actually rewrote these tests because they weren't very clear.